### PR TITLE
added aligner 

### PR DESCRIPTION
**Pull request by manmay-nakhashi**
this modules helps to learn phoneme alignment.

### CODE REVIEW

**Comment by manmay-nakhashi** [comment]
@lucidrains please review this pr , i have added phoneme aligner , compute pitch function from pytorch , losses for aligner and expand and combine function for durations and pitch.

File: naturalspeech2_pytorch/aligner.py [hunk 1]
```diff
@@ -0,0 +1,145 @@
+from typing import Tuple
+import torch
+from torch import nn
+import torch.nn.functional as F
+from einops import rearrange
+import numpy as np
+from numba import jit, prange
```
Comment (lucidrains) [inline naturalspeech2_pytorch/aligner.py:7]: hmm, was this here for the `maximum_path`?

File: naturalspeech2_pytorch/aligner.py [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+                padding=1,
+                bias=True,
+            ),
+            nn.ReLU(inplace=True),
+            nn.Conv1d(dim_in * 2, dim_in, kernel_size=1, padding=0, bias=True),
+            nn.ReLU(inplace=True),
+            nn.Conv1d(dim_in, attn_channels, kernel_size=1, padding=0, bias=True)
+        ])
+
+    def forward(self, queries: torch.Tensor, keys: torch.Tensor, mask: torch.Tensor = None):
```
Comment (lucidrains) [inline naturalspeech2_pytorch/aligner.py:49]: just my thinking out loud, but this is basically attention with the square of the euclidean distance

File: naturalspeech2_pytorch/aligner.py [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+            query_out = layer(query_out)
+
+        key_out = rearrange(key_out, 'b c t -> b c 1 t')
+        query_out = rearrange(query_out, 'b c t -> b c t 1')
+
+        attn_factor = (query_out - key_out) ** 2
+        attn_logp = -self.temperature * attn_factor.sum(1, keepdim=True)
+
+        if mask is not None:
+            attn_logp.data.masked_fill_(~mask.bool().unsqueeze(2), -float("inf"))
```
Comment (lucidrains) [inline naturalspeech2_pytorch/aligner.py:65]: unsqueeze can be replaced with einops

Comment (lucidrains) [reply naturalspeech2_pytorch/aligner.py:65]: i've also faced issues with using `inf` in the past (not sure if the pytorch team fixed that)

i like using `-torch.finfo(attn_logp.dtype).max` for safe measure

File: naturalspeech2_pytorch/aligner.py [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+        self.ctc_loss = torch.nn.CTCLoss(zero_infinity=True)
+        self.blank_logprob = blank_logprob
+
+    def forward(self, attn_logprob, in_lens, out_lens):
+        key_lens = in_lens
+        query_lens = out_lens
+        max_key_len = attn_logprob.size(-1)
+
+        # Reorder input to [query_len, batch_size, key_len]
+        attn_logprob = attn_logprob.squeeze(1).permute(1, 0, 2)
```
Comment (lucidrains) [inline naturalspeech2_pytorch/aligner.py:116]: einops rearrange here for clarity

File: naturalspeech2_pytorch/aligner.py [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+        # Convert to log probabilities
+        # Note: Mask out probs beyond key_len
+        device = attn_logprob.device
+        attn_logprob.masked_fill_(torch.arange(max_key_len + 1, device=device, dtype=torch.long).view(1, 1, -1) > key_lens.view(1, -1, 1), -1e15)
+
+        attn_logprob = self.log_softmax(attn_logprob)
+
+        # Target sequences
+        target_seqs = torch.arange(1, max_key_len + 1, device=device, dtype=torch.long).unsqueeze(0)
+        target_seqs = target_seqs.expand(key_lens.numel(), -1)
```
Comment (lucidrains) [inline naturalspeech2_pytorch/aligner.py:130]: prefer repeat here over expand

File: naturalspeech2_pytorch/naturalspeech2_pytorch.py [hunk 6]
```diff
@@ -80,6 +81,47 @@ def forward(self, x):
...
+        self.temperature = temperature
+        self.aligner = AlignerNet(dim_in = self.dim_in, 
+                                  dim_hidden = self.dim_hidden,
+                                  attn_channels = self.attn_channels,
+                                  temperature = self.temperature)
+    def forward(self, x, x_mask, y, y_mask):
+        attn_mask = torch.unsqueeze(x_mask, -1) * torch.unsqueeze(y_mask, 2)
+        alignment_soft, alignment_logprob = self.aligner(y.transpose(1, 2), x.transpose(1, 2), x_mask, None)
+        alignment_mas = maximum_path(
+            alignment_soft.squeeze(1).transpose(1, 2).contiguous(), attn_mask.squeeze(1).contiguous()
```
Comment (lucidrains) [inline naturalspeech2_pytorch/naturalspeech2_pytorch.py:107]: rearrange here for clarity

File: naturalspeech2_pytorch/naturalspeech2_pytorch.py [hunk 7]
```diff
@@ -80,6 +81,47 @@ def forward(self, x):
...
+        )
+        alignment_hard = torch.sum(alignment_mas, -1).int()
+        alignment_soft = alignment_soft.squeeze(1).transpose(1, 2)
```
Comment (lucidrains) [inline naturalspeech2_pytorch/naturalspeech2_pytorch.py:110]: rearrange here for clarity

File: naturalspeech2_pytorch/naturalspeech2_pytorch.py [hunk 8]
```diff
@@ -1280,7 +1322,7 @@ def forward(
             loss_weight = maybe_clipped_snr / (snr + 1)
 
         loss =  (loss * loss_weight).mean()
-
+
```
Comment (lucidrains) [inline naturalspeech2_pytorch/naturalspeech2_pytorch.py:1325]: extra spaces

**Review by lucidrains — COMMENTED** [review]
thank you Manmay! i don't know much about how phoneme alignment is done, but i left a few comments, mostly cosmetics. will do some more reading tomorrow morning and get it merged by monday

could you leave a short code snipper for the `Aligner` in the comments? (that runs without errors)

**Comment by lucidrains** [comment]
Screenshot from 2023-05-20 08-38-21

im combing the paper for mentions of the Alignernet you brought in, but i can't seem to find it

how did you conclude it is being used?

File: naturalspeech2_pytorch/aligner.py (hunk 2 above)
Comment (lucidrains) [reply naturalspeech2_pytorch/aligner.py:49]: ahh i see, basically creating an alignment matrix for ctc loss got it

**Comment by manmay-nakhashi** [comment]
@lucidrains all non autoregrassive will need to use alignment to gide the speech generation from phonemes. as this is non autoregrassive i concluded that we need an aligner there are multiple ways to get the alignments.
1) from MFA montreal forced aligner
2) monotonic alignment search(this is also one other option)
3) Aligner network over here by nvidia which is under mit license (they claim and it is seen in practical  to have a better alignment when audio length gets too long).
Aligner network will be trained along with the tts training so it'll jointly learn durations.

File: naturalspeech2_pytorch/aligner.py (hunk 1 above)
Comment (manmay-nakhashi) [reply naturalspeech2_pytorch/aligner.py:7]: yes i'll  remove the import. we can add if needed

**Comment by lucidrains** [comment]
@manmay-nakhashi thank you for your explanation! i believe you are correct that NAR solutions will require alignment still. they also needed some aligned conditioning signal for [another NAR based TTS solution](https://arxiv.org/abs/2305.09636)

let's try to get this merged by next week's end very latest! (i'll read that nvidia paper tomorrow morning)

thank you for the PR! :pray:

**Comment by manmay-nakhashi** [comment]
@lucidrains can we merge this now ?

**Comment by lucidrains** [comment]
@manmay-nakhashi oh hey, yea, but i left a few comments that were not addressed

**Comment by manmay-nakhashi** [comment]
@lucidrains i have resolved all the comments

**Comment by lucidrains** [comment]
@manmay-nakhashi thanks! there's one more comment though, regarding lack of `sqrt` for the l2 distance. maybe i misunderstood something

could you also add a small test main script within the files that run without error?

ex.

```python
if __name__ == '__main__':
  align = AlignNet(...)
  align(mock_tensor)
```

File: naturalspeech2_pytorch/aligner.py [hunk 9]
```diff
@@ -0,0 +1,143 @@
...
+        self,
+        dim_in=80,
+        dim_hidden=512,
+        attn_channels=80,
+        temperature=0.0005,
+    ):
+        super().__init__()
+        self.temperature = temperature
+        self.softmax = torch.nn.Softmax(dim=3)
+        self.log_softmax = torch.nn.LogSoftmax(dim=3)
```
Comment (lucidrains) [inline naturalspeech2_pytorch/aligner.py:19]: hmm, is this `log_softmax` being used anywhere?

Comment (manmay-nakhashi) [reply naturalspeech2_pytorch/aligner.py:19]: not used will remove this

File: naturalspeech2_pytorch/aligner.py [hunk 10]
```diff
@@ -0,0 +1,143 @@
...
+        key_out = rearrange(key_out, 'b c t -> b c 1 t')
+        query_out = rearrange(query_out, 'b c t -> b c t 1')
+
+        attn_factor = (query_out - key_out) ** 2
+        attn_logp = -self.temperature * attn_factor.sum(1, keepdim=True)
+
+        if mask is not None:
+            attn_logp.data.masked_fill_(~rearrange(mask, 'b t -> b t ()'), -torch.finfo(attn_logp.dtype).max)
+
+        attn = self.softmax(attn_logp)
```
[the remark this answers is not in the document]
Comment (lucidrains) [reply naturalspeech2_pytorch/aligner.py:65]: @manmay-nakhashi i think you need a square root after the sum? shouldn't one usually divide by temperature too? the command would be [`torch.cdist`](https://pytorch.org/docs/stable/generated/torch.cdist.html)

[the remark this answers is not in the document]
Comment (manmay-nakhashi) [reply naturalspeech2_pytorch/aligner.py:65]: @lucidrains you are right ill use torch.cdist

**Comment by manmay-nakhashi** [comment]
@lucidrains pr is ready tell me if i need to change anyting

**Comment by lucidrains** [comment]
@manmay-nakhashi thanks for the continuous polish

could you possibly remove all the print statements? also, i'm wondering why the maximum path function could not be done on the GPU?

**Comment by manmay-nakhashi** [comment]
@lucidrains fixed

**Comment by lexkoro** [comment]
@manmay-nakhashi In coqui tts there is this comment 
https://github.com/coqui-ai/TTS/blob/23a7a9a3633ee00e5bcd329d5f15b8c3f8971f8d/TTS/tts/utils/helpers.py#LL199C5-L199C74
but I haven't evaluated it myself.

**Comment by manmay-nakhashi** [comment]
@lexkoro i'll run the dummy loop and check

**Comment by manmay-nakhashi** [comment]
@lexkoro @lucidrains numpy is faster complete 1000 itr in 19 and torch completes it in 29 , i think because of gardient tracking :thinking:

**Comment by lucidrains** [comment]
@manmay-nakhashi if the hard path does not need gradients, you can wrap the function with the `torch.no_grad` decorator

**Comment by manmay-nakhashi** [comment]
@lucidrains with torch.no_grad() i am having limited speed bump still not as fast as numpy. 1000 itr. came down to 24 sec.

**Comment by manmay-nakhashi** [comment]
@lucidrains don't need to use binary loss only aligner loss and duration loss.

**Comment by lucidrains** [comment]
@manmay-nakhashi could you link to the line of code you are looking at?

**Comment by manmay-nakhashi** [comment]
https://github.com/lucidrains/naturalspeech2-pytorch/blob/a9b168341d5c87038eece5e6286bb560a0073381/naturalspeech2_pytorch/aligner.py#L144

**Comment by lucidrains** [comment]
@manmay-nakhashi got it, thanks!
